File: Loenn/entities/voidPortal.lua
local utils = require("utils")
local drawableSprite = require("structs.drawable_sprite")

local voidPortal = {}

voidPortal.name = "MaggyHelper/VoidPortal"
voidPortal.depth = -100
voidPortal.placements = {
    {
        name = "normal",
        data = {
            destinationId = "",
            isFinalPortal = false
        }
    },
    {
        name = "final",
        data = {
            destinationId = "",
            isFinalPortal = true
        }
    }
}

function voidPortal.sprite(room, entity)
    local sprite = drawableSprite.fromTexture("objects/void_portal/inactive", entity)
    sprite:setPosition(entity.x, entity.y)
    sprite:setJustification(0.5, 1.0)
    sprite.color = {0.5, 0.0, 0.5, 0.8}
    return sprite
end

function voidPortal.selection(room, entity)
    return utils.rectangle(entity.x - 24, entity.y - 64, 48, 64)
end

return voidPortal

File: Ahorn/entities/VoidPortal.jl
module VoidPortal

using ..Ahorn, Maple

@mapdef Entity "MaggyHelper/VoidPortal" VoidPortal(x::Integer, y::Integer, destinationId::String="", isFinalPortal::Bool=false)

const placements = Ahorn.PlacementDict(
    "normal" => Ahorn.EntityPlacement(VoidPortal),
    "final" => Ahorn.EntityPlacement(VoidPortal)
)

function Ahorn.selection(entity::VoidPortal)
    return Ahorn.Rectangle(entity.x - 8, entity.y - 8, 16, 16)
end

function Ahorn.render(ctx::Ahorn.Cairo.CairoContext, entity::VoidPortal, room::Maple.Room)
    Ahorn.drawRectangle(ctx, Ahorn.Rectangle(entity.x - 8, entity.y - 8, 16, 16), (0.5, 0.5, 0.5, 0.5), (0.0, 0.0, 0.0, 1.0))
end

end
